metrics/cognitive_complexity.py

```diff
@@ -0,0 +1,122 @@
+import tree_sitter
+
+from pathlib import Path
+from typing import List
+
+from byoqm.metric.metric import Metric
+from byoqm.metric.result import Result
+from byoqm.metric.violation import Violation
+from byoqm.source_repository.query_translations import translate_to
+from byoqm.source_repository.source_repository import SourceRepository
+
+
+class CognitiveComplexity(Metric):
+    def __init__(self):
+        self._source_repository: SourceRepository | None = None
+
+    def run(self) -> List:
+        """
+        A metric that counts the recursion and breaks in linear flow.
+
+        This metric is function/method specific
+        """
+        result = Result("cognitive complexity", [])
+        for file_path in self._source_repository.src_paths:
+            result.violations.extend(self._count_cognitive_complexity(file_path))
+        return result
+
+    def _count_cognitive_complexity(self, file_path: Path):
+        violations = []
+        ast: tree_sitter.Tree = self._source_repository.getAst(file_path)
+        lang = self._source_repository.language
+        initial_nodes_query_str = f"""{translate_to[lang]["function"]} @func"""
+        if lang == "c_sharp" or lang == "java":
+            initial_nodes_query_str += f"""{translate_to[lang]["constructor"]} @cons"""
```

Comment:
As far as i can see this is a concatenation of queries on the same string but is very specific, and would require us to add more checks to this if statement if other languages adhere to same semantics as c# or java.
Couldn't we just swap out two (unfortunately) large queries depending on language, where for c# and java it is the junction of both of these queries?

Reply:
After a group discussion, we decided to stick with this and accept that Nested Control Flow does not fit this standard